src/api/index.ts

```diff
@@ -31,17 +39,42 @@ export class ArgoCDApiClient implements ArgoCDApi {
     return `${proxyUrl}${this.proxyPath}`;
   }
 
+  private async fetchDecode<A, O, I>(url: string, typeCodec: t.Type<A, O, I>) {
+    const response = await fetch(url);
+    if (!response.ok) {
+      throw new Error(
+        `failed to fetch data, status ${response.status}: ${response.statusText}`,
+      );
+    }
+    const json = await response.json();
+    try {
+      return await tPromise.decode(typeCodec, json);
```

Comment:
We don't have to use `io-ts-promise` wrapper library, but then instead of promise API, we would deal with functional programming patterns - either monad, pipe and fold methods etc. It's fine for me, but could be hard for future contributors.

Reply:
I'm fine with either way. Probably easier for open source community to go with more known keywords.